### PR TITLE
Use response files on Windows

### PR DESCRIPTION
**Pull request by jimon**
If you use cmake + make than currently cmake does use response files when calling emcc. But when emcc tries to call clang - all arguments are passed as is, and we are hitting 32kb (?) limitation on Windows.

Proposed workaround - just always use WindowsPopen, and ask it to use response files with clang / llvm binaries.

### CODE REVIEW

**Comment by asRIA** [comment]
Can you take a look @juj ?

**Comment by juj** [comment]
This unconditionally enables the PIPE workaround for all Windows. That is not good, we should retain the  pipeing workaround as a specific option.

Agreed that we should run through response files when possible, and also have the feature run identically on all platforms, to avoid introducing diverting Windows-specific behavior.

It would be better to have the caller specify when to use response files instead of having to regex from the command filename. Perhaps we should make this a `EmPopen` object and prefer that over raw `Popen` wherever possible. @kripken : how does that sound to you?

**Comment by kripken** [comment]
Are you suggesting that everywhere in the codebase we would use `EmPopen` instead of the python standard `Popen`? Or just in some specific area?

**Comment by juj** [comment]
I think pretty much everywhere we'd use it - or at least in `tools/shared.py`, when we know we need response files, instead having to have each call site do their own response site crafting. The PIPEing hack would remain Windows only.

With `EmPopen` we could also consolidate the handling of `EM_BUILD_VERBOSE` to allow tracing through external processes that Emscripten calls, which is now restricted to a few hot places only.

**Comment by kripken** [comment]
I don't like avoiding the use of standard Python functionality, it makes the code less pythonic/idiomatic.

What would be the benefits from the change?

**Comment by jimon** [comment]
In our case we are most of all interested to use response files when calling clang compilation and linking steps, looks like at least first one is covered by https://github.com/kripken/emscripten/blob/master/tools/shared.py#L1944-L1951 so maybe we can do the thing there instead. It's pretty much very important, not sure how you folks missed it, on linux-like platforms there are ~2mb limitation for total command length, on Windows there is usually 32kb/8kb limitation.

Basically emcc response file functionality is broken now completely: sure you can use response file to pass arguments to emcc, but when emcc tries to pass them to clang - they are passed as is, which renders initial response file completely useless :(

**Comment by juj** [comment]
@kripken: Are you referring to the syntax? I.e. would you feel ok if this `EmPopen` was renamed to take over `Popen` instead?

There are three features we'd like to have over standard `Popen` in all places that call `Popen`:
   - use response files whenever invoking tools that support them
   - enable the use of the PIPE hack on Windows if needed ("if stdout or stderr is PIPEd, so must the other be")
   - enable tracing `Popen` calls via setting `EM_BUILD_VERBOSE=1/2/3`

**Comment by kripken** [comment]
Yes, the syntax is what worries me.

Surely we need response files in just a few places, though, not everywhere? There are very few commands that can have thousands of inputs (link and opt, basically)?

But if we can do it without new syntax, that sounds ok.
